### ISSUE TITLE
Add non-reversible parallel tempering

### ISSUE THREAD

**Issue by pawel-czyz** [blackjax-devs/blackjax#740]
### Presentation of the new sampler

Parallel tempering, known also as replica exchange MCMC, maintains $K$ Markov chains at different temperatures, ranging from $\pi_0$ (the reference distribution, for example the prior) to the target distribution $\pi =: \pi_{K-1}$.
Apart from local exploration kernels, targeting each distribution individually, it includes swap kernels, trying to switch states from different chains, hence targeting the distribution 
$$(x_0, \dotsc, x_{K-1})\mapsto \prod_{i=0}^{K-1} \pi_i(x_i),$$
defined on the spaces $\mathcal X^K = \mathcal X\times \cdots \mathcal X$. By retaining the samples from only the last coordinate, it allows one to sample from $\pi = \pi_{K-1}$.

Similarly to sequential Monte Carlo (SMC) samplers, this strategy can be highly efficient to sample from multimodal posteriors. Modern variant of parallel tempering, called non-reversible parallel tempering (NRPT), achieves the state-of-the-art performance in sampling from complex high-dimensional distributions. NRPT works with both discrete and continuous spaces, and allows one to leverage preliminary runs to tune the tempering schedule.

**Resources**
  - NRPT was introduced in [S. Syed et al., *Non-Reversible Parallel Tempering: a Scalable Highly Parallel MCMC Scheme*](https://doi.org/10.1111/rssb.12464). There is also an [excellent lecture](https://youtu.be/9gOssrhN3EA?feature=shared) available on this topic.
  - In Julia, NRPT is implemented in Pigeons.jl (see e.g., the [documentation](https://pigeons.run/stable/) or the [repository](https://github.com/Julia-Tempering/Pigeons.jl)).
  - I wrote [a blog post with a prototype of JAX implementation](https://pawel-czyz.github.io/posts/non-reversible-parallel-tempering.html), explaining the intuition behind the algorithm. 

### How does it compare to other algorithms in blackjax?

Compared with a single-chain MCMC sampler:
  - It requires maintaining $K$ chains. Hence, it can generally require more evaluations for the same number of samples collected.
  - However, it can lead to much faster mixing (especially for complex geometries or multimodal posteriors), resulting in a high effective sample size.
  - Computations for different chains can be parallelized.
  - Similarly to single-chain MCMC methods, warm-up samples have to be discarded. 

Compared with a tempered SMC sampler:
  - It maintains $K$ chains (each at its own temperature) over $T$ timesteps. SMC samplers keep all particles at the same temperature and then use a resampling step to move particles to the new temperature.
  - Similarly to SMC samplers, it works with arbitrary spaces (whether continuous or discrete), if efficient local exploration kernels (targeting the tempered distributions $\pi_i$) are available.
  - Similarly to some variants of tempered SMC samplers, non-reversible parallel tempering allows one to tune the tempering schedule basing on previous runs.
  - NRPT requires discarding the warm-up samples.
  
### Where does it fit in blackjax

BlackJAX offers a large collection of MCMC kernels. They can be leveraged to build non-reversible parallel tempering MCMC samplers, exploring different temperatures simultaneously, which leads to faster mixing and allows one to sample from multimodal posteriors. NRPT ac

### Are you willing to open a PR?

Yes. I have a [prototype implemented in a blog post](https://pawel-czyz.github.io/posts/non-reversible-parallel-tempering.html), which I would be willing to refactor and contribute.

I am however unsure about two design choices:

  1. Non-reversible parallel tempering requires application of local kernels, $K_i$, to individual coordinate chains $x_i \in \mathcal X$ (corresponding to tempered distributions $\pi_i$). What would be the best practice of storing and applying different kernels? For example, one may be willing to use HMC with a large step size for sampling from $\pi_0$ (or even sample from the prior directly, if e.g., using a probabilistic programming language) and HMC with a small step size to sample from $\pi=\pi_{K-1}$. In my prototype I employ kernels from the same family to use `jax.vmap`, rather than a `for` loop. I guess this problem is less apparent in tempered SMC samplers, where at temperature $T$ all particles are moved by the same kernel $K_T$. (Even though kernels $K_T$ may differ for different temperatures.)
  2. How to parallelize the computation? Pigeons.jl uses MPI communication between different machines to study very high-dimensional problems. Should a BlackJAX version use some version of sharding or could I keep it simple and rely on built-in parallelism?

**Comment by junpenglao** [comment]
Thank you for the detailed write-up, much appreciated. And yes, a contribution will be very welcome!

Regarding the design choice, `jax.vmap` would be the answer to both of your question.

I have not read in detail of your blog post, but just wondering if you have compared your implementation with [TFP](https://github.com/tensorflow/probability/blob/main/tensorflow_probability/python/mcmc/replica_exchange_mc.py), which i am a bit more familiar with.

**Comment by junpenglao** [comment]
BTW, I am a huge fan of parallel tempering - very excited about this! Looking forward to your PR!

**Comment by AdrienCorenflos** [comment]
Same here, I had planned to do it at some point but I've not been able to commit the time to :D very happy someone is doing it!

Design choice wise, I actually do not think it would be a good idea to vmap everything at the lower level, in particular in sight of being able to do proper sharding.
IMO there are two components to the method: 1) swap kernel, 2) reversible vs non-reversible application (this can probably be vmapped as it's onlt a chain on the indices conditionally on the log likelihood values).

Once that's done, the choice of parallelism for the state chains is very much user driven and it's hard to enforce a coherent interface supporting all JAX models.

**Comment by junpenglao** [comment]
IIUC, swapping kernel is basically swapping parameter (eg. step size), which means you update the input parameter with some advance indexing. The base kernel would remain the same like `step = jax.vmap(kernel.step)`

**Comment by junpenglao** [comment]
I think for simplicity, let's start with building the functionality assuming we are using the same base kernel (e.g., HMC) with different parameter (e.g., step_size)
